drivers/hwinfo/hwinfo_sam4l.c

```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright (c) 2020 Gerson Fernando Budke <nandojve@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <device.h>
+#include <drivers/hwinfo.h>
+#include <init.h>
+#include <soc.h>
+#include <string.h>
+
+#define MAX_SIZE_UID    15
+#define UID_ADDR        0x0080020C
+
+ssize_t z_impl_hwinfo_get_device_id(uint8_t *buffer, size_t length)
+{
+	size_t len = (length > MAX_SIZE_UID) ? MAX_SIZE_UID : length;
+	uint8_t *flash = (uint8_t *) UID_ADDR;
+	int i;
+
+	/* Copy the 120-bit unique ID. We cannot use memcpy as it would
+	 * execute code from flash.
+	 */
+	for (i = 0; i < len; i++) {
```

Comment:
a memcpy could be used here. Do you know if we should change the byte order? The datasheet does not give any information on the byte order. What does the asf do here?

Reply:
From the datasheet:
"Each device has a unique 120 bits serial number readable from address 0x0080020C to0x0080021A."

Reply:
The docs are not so good.

ASF sample do this:

```
	volatile uint8_t* unique_id_addr = (volatile uint8_t*) UNIQUE_ID_ADDR;

	/* Read the unique id and print it */
	for (i=0; i<15; i++) {
		printf("%02x",*unique_id_addr++);
	}
```
SAM-BA bootloader shows:

```
(sam-ba_2.18) 1 % FLASH::ReadUniqueSerialNumber
buffer address 0x20002a88
Unique Serial Number :
	0x00 : 0x1c
	0x01 : 0xda
	0x02 : 0xa0
	0x03 : 0xc3
	0x04 : 0x62
	0x05 : 0x9f
	0x06 : 0xec
	0x07 : 0xd3
	0x08 : 0x12
	0x09 : 0xd4
	0x0a : 0x9c
	0x0b : 0xbf
	0x0c : 0x6c
	0x0d : 0xdf
	0x0e : 0xec
(sam-ba_2.18) 1 % 
```

I follow same byte order than SAM4/V7 already uses. This way the lower address 0x0080020C will be print as most significant by Zephyr as below:

uart:~$ hwinfo devid 
Length: 15
ID: 0x1cdaa0c3629fecd312d49cbf6cdfec

Reply:
So lets use memcpy and we are good to go.

Reply:
github is taking time to update files...don't know why. The hash that I sent with memcpy had been showing old version by ~25min. Right now I can see the right version...wierd!

Reply:
Do we need the length < 0. size_t is unsigned anyway.